### DISCUSSION TITLE
panic: runtime error: slice bounds out of range [:12826251916321229408] with capacity 884736

### DISCUSSION THREAD

**Discussion by ceccocats** [foxglove/mcap#1173] [General]
Hi, 
I have a corrupted mcap that I cannot read. 
If i run `mcap info` it works like expected, but any other commands that actually read data don't work:
```
ceccocats@aron:~/Downloads$ mcap cat  20240427_yasgp_qualifiers_3_1_raw.tkmcap 
panic: runtime error: slice bounds out of range [:12826251916321229408] with capacity 884736

goroutine 1 [running]:
github.com/foxglove/mcap/go/mcap.(*indexedMessageIterator).Next(0xc00051e0e0, {0x43faa5?, 0xc0000021a0?, 0x300000002?})
	/home/runner/work/mcap/mcap/go/mcap/indexed_message_iterator.go:239 +0x2ea
github.com/foxglove/mcap/go/cli/mcap/cmd.printMessages({0xc0002efc00?, 0xc0002ff480?}, {0x10253e0, 0xc000010018}, {0x1022300, 0xc00051e0e0}, 0x0)
	/home/runner/work/mcap/mcap/go/cli/mcap/cmd/cat.go:97 +0x1e5
github.com/foxglove/mcap/go/cli/mcap/cmd.glob..func5.1(0x1d?, {0x7f418470c838?, 0xc000010ae8})
	/home/runner/work/mcap/mcap/go/cli/mcap/cmd/cat.go:222 +0x1c5
github.com/foxglove/mcap/go/cli/mcap/utils.WithReader({0x102a410, 0xc000036088}, {0x7ffece60621d?, 0x0?}, 0xc00025fd90)
	/home/runner/work/mcap/mcap/go/cli/mcap/utils/utils.go:104 +0x431
github.com/foxglove/mcap/go/cli/mcap/cmd.glob..func5(0x1549240?, {0xc0002ff250, 0x1, 0x1?})
	/home/runner/work/mcap/mcap/go/cli/mcap/cmd/cat.go:212 +0x312
github.com/spf13/cobra.(*Command).execute(0x1549240, {0xc0002ff220, 0x1, 0x1})
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:876 +0x67b
github.com/spf13/cobra.(*Command).ExecuteC(0x1548340)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:990 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:918
github.com/foxglove/mcap/go/cli/mcap/cmd.Execute()
	/home/runner/work/mcap/mcap/go/cli/mcap/cmd/root.go:21 +0x25
main.main()
	/home/runner/work/mcap/mcap/go/cli/mcap/main.go:6 +0x17
``` 
```
ceccocats@aron:~/Downloads$ mcap doctor  20240427_yasgp_qualifiers_3_1_raw.tkmcap
Header.profile field "tk" is not a well-known profile.
Uncompressed chunk data size != Chunk.uncompressed_size
Got a Message record for channel: 48 before a channel info.
Got a Message record for channel: 47 before a channel info.
Got a Message record for channel: 52 before a channel info.
[...]
```

Do you know what can be the issue that generated this corrupted file?

you can find the corrupted mcap here: https://drive.google.com/file/d/1LU6N1_wUwoZfJ5E9I26zqScQjcgVGcb4/view?usp=drive_link

**Comment by jtbandes** [comment]
The download link you shared doesn't work. Can you zip the mcap file and drag & drop it here into the comment field?

To answer your question directly:

> Do you know what can be the issue that generated this corrupted file?

It seems like the file is corrupt. `Uncompressed chunk data size != Chunk.uncompressed_size` indicates that a chunk was decompressed and it ended up at a different size than expected, which means the file is not valid and was either written incorrectly or was later corrupted. How did you create this file?

**Comment by ceccocats** [comment]
Unfortunately the mcap Is big (300mb) so it can't be uploaded here, I updated the link now it should be accessible:
[drive.google.com/file/d/1LU6N1_wUwoZfJ5E9I26zqScQjcgVGcb4/view?usp=drive_link](https://drive.google.com/file/d/1LU6N1_wUwoZfJ5E9I26zqScQjcgVGcb4/view?usp=drive_link)

We recorded the mcap using mcap::McapWriter api in c++. It could be a racing condition in our code, but It would be wonderful if the file Is somehow recoverable.
Mcap recover gives this error:
```
ceccocats@aron:~/Downloads$ mcap recover 20240427_yasgp_qualifiers_3_1_raw.tkmcap -o recover.tkmcap
Recovered 0 messages, 0 attachments, and 0 metadata records.
failed to decompress chunk: unexpected EOF
```

**Reply by jtbandes** [reply]
> It could be a racing condition in our code

This sounds likely. The McapWriter is not thread safe. You need to synchronize all accesses to it.

**Reply by ceccocats** [reply]
you know if there is a possibility to recover this file? thanks
